dpcpp/matrix/batch_csr_kernels.hpp.inc

```diff
@@ -0,0 +1,84 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+template <typename ValueType, typename IndexType>
+__dpct_inline__ void simple_apply_kernel(
+    const gko::batch::matrix::csr::batch_item<const ValueType, IndexType>& mat,
+    const ValueType* b, ValueType* x, sycl::nd_item<3>& item_ct1)
+{
+    const auto num_rows = mat.num_rows;
+    const auto val = mat.values;
+    const auto col = mat.col_idxs;
+    for (int tidx = item_ct1.get_local_linear_id(); tidx < num_rows;
+         tidx += item_ct1.get_local_range().size()) {
+        auto temp = zero<ValueType>();
+        for (auto nnz = mat.row_ptrs[tidx]; nnz < mat.row_ptrs[tidx + 1];
+             nnz++) {
+            const auto col_idx = col[nnz];
+            temp += val[nnz] * b[col_idx];
+        }
+        c[tidx] = temp;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__dpct_inline__ void advanced_apply_kernel(
+    const ValueType alpha,
+    const gko::batch::matrix::csr::batch_item<const ValueType, IndexType>& mat,
+    const ValueType* b, const ValueType beta, ValueType* x,
+    sycl::nd_item<3>& item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+
+    for (int row = sg.get_group_id(); row < mat.num_rows;
+         row += sg.get_group_range().size()) {
+        const int row_start = mat.row_ptrs[row];
+        const int row_end = mat.row_ptrs[row + 1];
+
+        ValueType temp = zero<ValueType>();
+        for (int i = sg.get_local_id() + row_start; i < row_end;
+             i += sg.get_local_range().size()) {
+            const int col = mat.col_idxs[i];
+            const ValueType val = mat.values[i];
+            temp += val * b[col];
+        }
+
+        temp = ::gko::kernels::dpcpp::reduce(
+            sg, temp, [](ValueType a, ValueType b) { return a + b; });
+
+
+        if (sg.get_local_id() == 0) {
+            c[row] = alpha * temp + beta * c[row];
+        }
+    }
```

Comment:
This is a different level of optimization than the other kernels. Is that intended?